### PR TITLE
Verify Wheel

### PR DESCRIPTION
**Pull request by dkeeney**
The intent of the PR is to verify that the wheel package contains the right things.

For #19

### CODE REVIEW

**Comment by dkeeney** [comment]
This is work in progress.
I found that the wheel did not contain the py unit tests.  So I set about to copy the unit tests into directory where the wheel package is created (build/Release/distr/src/tests)

This involved:
- moving setup.py from bindings/py/packaging to the top directory and eliminating the redirection call it was making.  So the requirements.txt and the real setup.py are now both at the top of the repository.
- moving the bindings unit tests (bindings/py/tests) to bindings/py/packaging/src/tests where it could be parallel with htm.  Copying bindings/py/packaging/ to the distr folder  gets both htm and tests folders with one copy.
- copy py/tests to the distr folder (merging with the tests from packaging).
- both copies had to filter out the __pycache__ files.

That is as far as I have gotten so far.  The next step is to build it  and create the wheel like this for all platforms and see if everything is there in the wheel and that it can be installed using pip directly from the wheel.

Then I need to deal with the linux binary distribution thing.

File: bindings/py/README.md [hunk 1]
```diff
@@ -1,33 +1,23 @@
 
 ## Layout of directories:
 ```
-  Repository
-     bindings
+  REPO_DIR    external  -- cmake scripts to download/build dependancies
```
Comment (breznak) [inline bindings/py/README.md]: nit, looks like your editor adds weird line ends

File: setup.py-old [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# ----------------------------------------------------------------------
+# HTM Community Edition of NuPIC
```
Comment (breznak) [inline setup.py-old]: this is just a backup (?) I assume it'll go away before merge

**Comment by breznak** [comment]
Failing cmake call
https://github.com/htm-community/htm.core/pull/694/checks?check_run_id=238251520

**Comment by dkeeney** [comment]
No, I have lots of work to do on this before it is ready to merge.

**Comment by dkeeney** [comment]
Working on Windows....
- I build with `python setup.py install --user --force`
- I build the wheel with `python setup.py bdist_wheel`
- Use pip to uninstall  `python -m pip uninstall htm.core`
- Use pip to install from wheel  `python -m pip install build/Release/distr/dist/htm.core-2.0.14-cp37-cp37m-win_amd64.whl`

It says it installs `Successfully installed htm.core-2.0.14`
but I am getting this error:

```
ERROR: Error checking for conflicts.
Traceback (most recent call last):
  File "C:\Users\Dave\AppData\Local\Programs\Python\Python37\lib\site-packages\pip\_vendor\pkg_resources\__init__.py", line 3012, in _dep_map
    return self.__dep_map
  File "C:\Users\Dave\AppData\Local\Programs\Python\Python37\lib\site-packages\pip\_vendor\pkg_resources\__init__.py", line 2806, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During handling of the above exception, another exception occurred:
```

I have no idea what might be causing it.

**Comment by dkeeney** [comment]
Hmmm, I tried to install another unrelated package  and got the same error.  so maybe its something to do with the way I have python installed....and nothing to do with the wheel.

**Comment by breznak** [comment]
> Use pip to install from wheel python -m pip install build/Release/distr/dist/htm.core-2.0.14-cp37-cp37m-win_amd64.whl

hey, we could probably test this as a part of the CI steps! The wheel is there, we can just uninstall the source-built library, and try if installing from `whl` works (+run a few tests)

**Comment by breznak** [comment]
@dkeeney I added the CI check that the wheel is installable

**Comment by dkeeney** [comment]
After re-installing Python 3.7 and removing all other Pythons, I still get this error about conflicts.
But since it did say the install was successful.  So I decided to try to install all dependencies as well.
```
pip install --user --force build/Release/distr/dist/htm.core-2.0.14-cp37-cp37m-win_amd64.whl
```
It gave me that same error as before.  --ignoring that for now.
Then it gave me an error trying to install an egg .  The error was `cannot find the file specified.` and the file was   `hexy-1.4.3-py3.7.egg`
I can see that it downloaded and built the wheel for hexy  but perhaps what it is actually doing is installing something else that needs hexy and hexy is not there yet.

Anyway, the point is....this still does not work yet.

**Comment by dkeeney** [comment]
Also, I cannot figure out how to run the .py unit tests without the full repository being present.
I can get them into the wheel but don't know how to run them.

**Comment by dkeeney** [comment]
@breznak you added `python -m pip install --force build/Release/distr/dist/*.whl`  to the CI script.  On Linux that works but on windows, the * is not expanded.
```
J:\Projects\AI\htmG>python -m pip install build/Release/distr/dist/*.whl
Requirement 'build/Release/distr/dist/*.whl' looks like a filename, but the file does not exist
*.whl is not a valid wheel filename.
```
I verified that the wheel does exists.
`build/Release/distr/dist/htm.core-2.0.14-cp37-cp37m-win_amd64.whl`

**Comment by dkeeney** [comment]
Traced down the Conflict error.  It is a corruption in my site-packages directory caused by an earlier wheel install failure for which pip did not clean up correctly.   Anyway, that one we can safely ignore.

**Comment by dkeeney** [comment]
> In your CI script   `python -m pip install build/Release/distr/dist/*.whl`

I added '--force' so that it will also verify that the wheel has the right requirements.txt and that they all can be  installed.

That does seem to work here on windows if I expand the '*' before making the call.

**Comment by dkeeney** [comment]
> `python -m pytest bindings/py/packaging/tests`

If I change this to `python -m pytest bindings/py/tests`
it does work :-)

**Comment by breznak** [comment]
> Traced down the Conflict error. It is a corruption in my site-packages directory caused by an earlier wheel install failure for which pip did not clean up correctly. Anyway, that one we can safely ignore. 

glad you had it figured. For these reasons, I always use python virtualenv (`python -m venv`), and if things break, I just rm the whole install.

File: .github/workflows/htmcore.yml [hunk 3]
```diff
@@ -93,10 +93,12 @@ jobs:
         path: build/Release/distr/dist
 
     - name: Test python package is working
+      shell: bash
```
Comment (breznak) [inline .github/workflows/htmcore.yml]: nice, this forces windows to act compatible with unixes, right? (is it using the WSL?)

File: .github/workflows/htmcore.yml [hunk 4]
```diff
@@ -93,10 +93,12 @@ jobs:
...
       run: |
         python -m pip uninstall -y htm.core 
-        python -m pip install --force build/Release/distr/dist/*.whl
-        python -m pytest bindings/py/packaging/tests
+        x=$(ls build/Release/distr/dist/*.whl)
```
Comment (breznak) [inline .github/workflows/htmcore.yml]: nit: change x to fileName

**Comment by breznak** [comment]
> ls: build/Release/distr/dist/*.whl: No such file or directory 

oh, ok, this is an unrelated err. 

I've temporarily allowed some steps (that should be run only on tag/release) to run on each PR for debugging. But the creation of the whl is still only on tag.

File: bindings/py/README.md (hunk 1 above)
Comment (breznak) [reply bindings/py/README.md]: @dkeeney please fix these

File: setup.py-old (hunk 2 above)
Comment (breznak) [reply setup.py-old]: FIXME rm this file

Comment (breznak) [reply setup.py-old]: ok, it has been gone

**Comment by breznak** [comment]
```
Run ls dist/
htm.core-2.0.16-cp37-cp37m-linux_x86_64.whl
htm.core-2.0.16-cp37-cp37m-macosx_10_14_x86_64.whl
htm.core-2.0.16-cp37-cp37m-win_amd64.whl
htm.core-2.0.16-py3.7-linux-x86_64.egg
htm.core-2.0.16-py3.7-macosx-10.14-x86_64.egg
htm.core-2.0.16-py3.7-win-amd64.egg
requirements.txt
```

why is `egg` still being created?

**Comment by dkeeney** [comment]
> why is egg still being created?

It occurred to me in my sleep last night.  The default for the setup.py setup() function is to generate an egg when we use the keyword `install`.  So when we do `python setup.py install --user --force` it will create the egg.  But if we do
`python setup.py bdist_wheel --install --user --force` instead to do the build it will generate a wheel as a product of the full build.

I wanted to try this out this morning.

**Comment by dkeeney** [comment]
> python setup.py bdist_wheel --install --user --force

Oh, that should be `python setup.py bdist_wheel --user --force`

**Comment by dkeeney** [comment]
> Oh, that should be python setup.py bdist_wheel --user --force

Well.... it does not like --user --force, so the overall build command is now **`python setup.py bdist_wheel`**
What this will NOT do is install locally.  Apparently it is the local install that uses the .egg file.

The CI builds don't need the local egg install if we use the wheel to install.

**Comment by dkeeney** [comment]
My schedule:
I have some commitments this morning but I can work this after noon and Sunday.  But Monday we are traveling to Algodones Mexico for dental work.  May be gone 4-5 days.

**Comment by breznak** [comment]
> It occurred to me in my sleep last night.  

wow, nice :hash: do you dream of programming and electric sheep? :) 

> build command is now python setup.py bdist_wheel 

I'll try to use this.

**Comment by breznak** [comment]
Uploading linux wheels to PYPI works! 
https://github.com/htm-community/htm.core/pull/694/checks?check_run_id=246117912#step:8:15

**Comment by breznak** [comment]
Everything seems to work! Merging to test PYPI upload with a new tag

**Pull request merged**